File: cpp/core/src/zxing/common/reedsolomon/GenericGF.cpp
// -*- mode:c++; tab-width:2; indent-tabs-mode:nil; c-basic-offset:2 -*-
/*
 *  GenericGF.cpp
 *  zxing
 *
 *  Created by Lukas Stabe on 13/02/2012.
 *  Copyright 2012 ZXing authors All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "GenericGF.h"

#include <zxing/common/IllegalArgumentException.h>   // for IllegalArgumentException
#include <zxing/common/reedsolomon/GenericGFPoly.h>  // for GenericGFPoly

#include "zxing/common/Array.h"                      // for ArrayRef, Array
#include "zxing/common/Counted.h"                    // for Ref

#include <Utils/Macros.h>

using pping::GenericGF;
using pping::GenericGFPoly;
using pping::Ref;

Ref<GenericGF> GenericGF::QR_CODE_FIELD_256(new GenericGF(0x011D, 256, 0));
Ref<GenericGF> GenericGF::DATA_MATRIX_FIELD_256(new GenericGF(0x012D, 256, 1));
Ref<GenericGF> GenericGF::AZTEC_PARAM(new GenericGF(0x13, 16, 1));
Ref<GenericGF> GenericGF::AZTEC_DATA_6(new GenericGF(0x43, 64, 1));
Ref<GenericGF> GenericGF::AZTEC_DATA_8(GenericGF::DATA_MATRIX_FIELD_256);
Ref<GenericGF> GenericGF::AZTEC_DATA_10(new GenericGF(0x409, 1024, 1));
Ref<GenericGF> GenericGF::AZTEC_DATA_12(new GenericGF(0x1069, 4096, 1));
Ref<GenericGF> GenericGF::MAXICODE_FIELD_64(GenericGF::AZTEC_DATA_6);
  
namespace {
  int INITIALIZATION_THRESHOLD = 0;
}

/* Dont't use this directly, use the factory method! */
GenericGF::GenericGF(int primitive, int size, int b) MB_NOEXCEPT_EXCEPT_BADALLOC
  : size_(size), primitive_(primitive), generatorBase_(b), initialized_(false) {}

pping::FallibleRef<GenericGF> GenericGF::createGenericGF(int primitive, int size, int b) MB_NOEXCEPT_EXCEPT_BADALLOC
{
    auto genericGF(Ref<GenericGF>(new GenericGF(primitive, size, b)));

    if (size <= INITIALIZATION_THRESHOLD) {
        auto const init(genericGF->initialize());
        if(!init)
            return init.error();
    }
    return genericGF;
}
  
pping::Fallible<void> GenericGF::initialize() MB_NOEXCEPT_EXCEPT_BADALLOC {
  expTable_.resize(size_);
  logTable_.resize(size_);
    
  int x = 1;
    
  for (int i = 0; i < size_; i++) {
    expTable_[i] = x;
    x <<= 1; // x = x * 2; we're assuming the generator alpha is 2
    if (x >= size_) {
      x ^= primitive_;
      x &= size_-1;
    }
  }
  for (int i = 0; i < size_-1; i++) {
    logTable_[expTable_[i]] = i;
  }
  //logTable_[0] == 0 but this should never be used
    
  auto const tryCreateGenericPoly(GenericGFPoly::createGenericGFPoly(*this, ArrayRef<int>(new Array<int>(1))));
  if(!tryCreateGenericPoly)
      return tryCreateGenericPoly.error();

  zero_ = *tryCreateGenericPoly;
  zero_->getCoefficients()[0] = 0;

  auto const tryCreateGenericPolyOne(GenericGFPoly::createGenericGFPoly(*this, ArrayRef<int>(new Array<int>(1))));
  if(!tryCreateGenericPolyOne)
      return tryCreateGenericPolyOne.error();

  one_ = *tryCreateGenericPolyOne;
  one_->getCoefficients()[0] = 1;
    
  initialized_ = true;

  return success();
}
  
pping::Fallible<void> GenericGF::checkInit() MB_NOEXCEPT_EXCEPT_BADALLOC {
  if (!initialized_) {
    auto const init(initialize());
    if(!init)
        return init.error();
  }
  return success();
}
  
pping::FallibleRef<GenericGFPoly> GenericGF::getZero() MB_NOEXCEPT_EXCEPT_BADALLOC {
  auto const check(checkInit());
  if(!check)
      return check.error();
  return zero_;
}
  
pping::FallibleRef<GenericGFPoly> GenericGF::getOne() MB_NOEXCEPT_EXCEPT_BADALLOC {
    auto const check(checkInit());
    if(!check)
        return check.error();
    return one_;
}
  
pping::FallibleRef<GenericGFPoly> GenericGF::buildMonomial(int degree, int coefficient) MB_NOEXCEPT_EXCEPT_BADALLOC {
  auto const check(checkInit());
  if(!check)
    return check.error();
    
  if (degree < 0) {
    return failure<IllegalArgumentException>("Degree must be non-negative");
  }
  if (coefficient == 0) {
    return zero_;
  }
  ArrayRef<int> coefficients(new Array<int>(degree + 1));
  coefficients[0] = coefficient;
    
  return GenericGFPoly::createGenericGFPoly(*this, coefficients);
}
  
int GenericGF::addOrSubtract(int a, int b) {
  return a ^ b;
}
  
pping::Fallible<int> GenericGF::exp(int a) MB_NOEXCEPT_EXCEPT_BADALLOC {
  auto const check(checkInit());
  if(!check)
    return check.error();

  return expTable_[a];
}
  
pping::Fallible<int> GenericGF::log(int a) MB_NOEXCEPT_EXCEPT_BADALLOC {
    if(a == 0)
        return failure<IllegalArgumentException>("cannot give log(0)");

    auto const check(checkInit());
    if(!check)
      return check.error();

    return logTable_[a];
}
  
pping::Fallible<int> GenericGF::inverse(int a) MB_NOEXCEPT_EXCEPT_BADALLOC {
    if(a == 0)
        return failure<IllegalArgumentException>("Cannot calculate the inverse of 0");

    auto const check(checkInit());
    if(!check)
      return check.error();

    return expTable_[size_ - logTable_[a] - 1];
}
  
pping::Fallible<int> GenericGF::multiply(int a, int b) MB_NOEXCEPT_EXCEPT_BADALLOC {
  auto const check(checkInit());
  if(!check)
    return check.error();
    
  if (a == 0 || b == 0) {
    return 0;
  }
    
  return expTable_[(logTable_[a] + logTable_[b]) % (size_ - 1)];
  }
    
int GenericGF::getSize() {
  return size_;
}

int GenericGF::getGeneratorBase() {
  return generatorBase_;
}

File: core/src/main/java/com/google/zxing/common/reedsolomon/GenericGF.java
/*
 * Copyright 2007 ZXing authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.zxing.common.reedsolomon;

/**
 * <p>This class contains utility methods for performing mathematical operations over
 * the Galois Fields. Operations use a given primitive polynomial in calculations.</p>
 *
 * <p>Throughout this package, elements of the GF are represented as an {@code int}
 * for convenience and speed (but at the cost of memory).
 * </p>
 *
 * @author Sean Owen
 * @author David Olivier
 */
public final class GenericGF {

  public static final GenericGF AZTEC_DATA_12 = new GenericGF(0x1069, 4096, 1); // x^12 + x^6 + x^5 + x^3 + 1
  public static final GenericGF AZTEC_DATA_10 = new GenericGF(0x409, 1024, 1); // x^10 + x^3 + 1
  public static final GenericGF AZTEC_DATA_6 = new GenericGF(0x43, 64, 1); // x^6 + x + 1
  public static final GenericGF AZTEC_PARAM = new GenericGF(0x13, 16, 1); // x^4 + x + 1
  public static final GenericGF QR_CODE_FIELD_256 = new GenericGF(0x011D, 256, 0); // x^8 + x^4 + x^3 + x^2 + 1
  public static final GenericGF DATA_MATRIX_FIELD_256 = new GenericGF(0x012D, 256, 1); // x^8 + x^5 + x^3 + x^2 + 1
  public static final GenericGF AZTEC_DATA_8 = DATA_MATRIX_FIELD_256;
  public static final GenericGF MAXICODE_FIELD_64 = AZTEC_DATA_6;

  private final int[] expTable;
  private final int[] logTable;
  private final GenericGFPoly zero;
  private final GenericGFPoly one;
  private final int size;
  private final int primitive;
  private final int generatorBase;

  /**
   * Create a representation of GF(size) using the given primitive polynomial.
   *
   * @param primitive irreducible polynomial whose coefficients are represented by
   *  the bits of an int, where the least-significant bit represents the constant
   *  coefficient
   * @param size the size of the field
   * @param b the factor b in the generator polynomial can be 0- or 1-based
   *  (g(x) = (x+a^b)(x+a^(b+1))...(x+a^(b+2t-1))).
   *  In most cases it should be 1, but for QR code it is 0.
   */
  public GenericGF(int primitive, int size, int b) {
    this.primitive = primitive;
    this.size = size;
    this.generatorBase = b;

    expTable = new int[size];
    logTable = new int[size];
    int x = 1;
    for (int i = 0; i < size; i++) {
      expTable[i] = x;
      x *= 2; // we're assuming the generator alpha is 2
      if (x >= size) {
        x ^= primitive;
        x &= size - 1;
      }
    }
    for (int i = 0; i < size - 1; i++) {
      logTable[expTable[i]] = i;
    }
    // logTable[0] == 0 but this should never be used
    zero = new GenericGFPoly(this, new int[]{0});
    one = new GenericGFPoly(this, new int[]{1});
  }

  GenericGFPoly getZero() {
    return zero;
  }

  GenericGFPoly getOne() {
    return one;
  }

  /**
   * @return the monomial representing coefficient * x^degree
   */
  GenericGFPoly buildMonomial(int degree, int coefficient) {
    if (degree < 0) {
      throw new IllegalArgumentException();
    }
    if (coefficient == 0) {
      return zero;
    }
    int[] coefficients = new int[degree + 1];
    coefficients[0] = coefficient;
    return new GenericGFPoly(this, coefficients);
  }

  /**
   * Implements both addition and subtraction -- they are the same in GF(size).
   *
   * @return sum/difference of a and b
   */
  static int addOrSubtract(int a, int b) {
    return a ^ b;
  }

  /**
   * @return 2 to the power of a in GF(size)
   */
  int exp(int a) {
    return expTable[a];
  }

  /**
   * @return base 2 log of a in GF(size)
   */
  int log(int a) {
    if (a == 0) {
      throw new IllegalArgumentException();
    }
    return logTable[a];
  }

  /**
   * @return multiplicative inverse of a
   */
  int inverse(int a) {
    if (a == 0) {
      throw new ArithmeticException();
    }
    return expTable[size - logTable[a] - 1];
  }

  /**
   * @return product of a and b in GF(size)
   */
  int multiply(int a, int b) {
    if (a == 0 || b == 0) {
      return 0;
    }
    return expTable[(logTable[a] + logTable[b]) % (size - 1)];
  }

  public int getSize() {
    return size;
  }
  
  public int getGeneratorBase() {
    return generatorBase;
  }
  
  @Override
  public String toString() {
    return "GF(0x" + Integer.toHexString(primitive) + ',' + size + ')';
  }
  
}
